Original patch was:
http://sourceware.org/ml/gdb-patches/2007-12/msg00397.html
http://sourceware.org/ml/gdb-cvs/2007-12/msg00123.html

extended for the RHEL safety:

2007-12-28  Jan Kratochvil  <jan.kratochvil@redhat.com>

	* dwarf2read.c (fixup_partial_die): Provide full
	`DW_TAG_class_type'-type backing for `DW_TAG_interface_type', even for
	namespaces which should not apply for Java `DW_TAG_interface_type'.

2008-02-24  Jan Kratochvil  <jan.kratochvil@redhat.com>

	Port to GDB-6.8pre.

Index: ./gdb/dwarf2read.c
===================================================================
RCS file: /cvs/src/src/gdb/dwarf2read.c,v
retrieving revision 1.245
diff -u -p -r1.245 dwarf2read.c
--- ./gdb/dwarf2read.c	26 Dec 2007 12:36:18 -0000	1.245
+++ ./gdb/dwarf2read.c	27 Dec 2007 23:25:49 -0000
@@ -5887,7 +5887,8 @@ fixup_partial_die (struct partial_die_in
 
   /* Set default names for some unnamed DIEs.  */
   if (part_die->name == NULL && (part_die->tag == DW_TAG_structure_type
-				 || part_die->tag == DW_TAG_class_type))
+				 || part_die->tag == DW_TAG_class_type
+				 || part_die->tag == DW_TAG_interface_type))
     part_die->name = "(anonymous class)";
 
   if (part_die->name == NULL && part_die->tag == DW_TAG_namespace)
@@ -5895,6 +5896,7 @@ fixup_partial_die (struct partial_die_in
 
   if (part_die->tag == DW_TAG_structure_type
       || part_die->tag == DW_TAG_class_type
+      || part_die->tag == DW_TAG_interface_type
       || part_die->tag == DW_TAG_union_type)
     guess_structure_name (part_die, cu);
 }
